### ISSUE TITLE
WebMap Plugin: Instantiation and Projection

### ISSUE THREAD

**Issue by sindizzy** [DotSpatial/DotSpatial#1296]
DotSpatial version: 2.0
So I am trying to integrate a web service into my map and I know the WebMap plugin has the functionality. So I have a couple questions:

1. Is there a way to add a web service to my map without using the AppManager? From the looks of it I think I have to let the AppManager add the extensions. But then it creates default controls. That is fine but I use a heavily customized GUI with customized Map, Legend, buttons , tools, etc. I end up having to go through hoops and ladders to hide the default controls so I can use my own controls. Then some hacks to tap into the Service Provider selector. It works but very hacky. Wanted to make sure there is no other way to do this without the overhead of the AppManager.

2. Looking at the source code it appears that when the web service is consumed it automatically sets the map projection to Mercator. It also reprojects all layers to that projection. Seems to work but now some of my other code gets hosed. Specifically the DynamicVisibility gets broken because the width should now be in meters instead of degrees. Some other issues too due to the reprojection. Is there any hope of reprojecting the images to the current map projection instead of the conforming to the projection of the images?

~AGP

**Comment by VectorZita** [comment]
Using the specific plugin functionality as "not a plugin" is definitely possible but, of course, it is going to take some refactoring, i.e. reconstructing the project to only isolate the objects of interest.

Furthermore, if adapting your tools to the Mercator (Web Mercator actually) projection is totally out of the question, it is going to take some more effort to have it the other way around (i.e. project everything to the map projection).

**Comment by sindizzy** [comment]
I will take a look at some of the code this weekend.

**Comment by sindizzy** [comment]
So I am experimenting with adding geo images (gid +gfw) and then reprojecting the map. It looks like images can be reprojected. So with this in mind the web maps end up being rendered as images but cant quite understand why they cant be reprojected. I will do some more testing.

**Comment by VectorZita** [comment]
Images can certainly be reprojected but it is not as simple as for points. It is also not as simple as it is for images as well... Because map projections always affect the underlying geometry, reprojecting pixels as points typically ends up creating an image with blanks in-between. Due to that fact, image reprojection is implemented in a "lightweight" manner in `DotSpatial` (if I don't miss my guess, that is). 

From what I have seen, the standard raster `DataSet` implementation in `DotSpatial` is based on the `RasterBoundDataSet` abstract base class and `MapImageLayer`s can be built on top implementations thereof. Out of what I have seen, no class derived from `RasterBoundDataSet` actually overrides the `Reproject()` method. Checking the methodology, I find that the base class reprojects the image data using an Affine Transformation.

The thing is, affine transformations are far too simple to serve as approximations of geodetic map projections. So, are you adding the images as `MapImageLayer`s? Are you absolutely certain that the images are reprojecting properly? Did you try overlaying images with other loaded shapes to see if they are OK? Did you check this on a relatively larger scale (i.e. zoomed out)?

**Comment by sindizzy** [comment]
Yes, after looking at the code yesterday the only thing I saw was AffineTransformation attached to image reprojection, just as you have stated.

In my test I added a raster of terrain, then a polygon layer for the US states, then a gif image of weather radar.

![image](https://user-images.githubusercontent.com/5298460/63654942-1a569c80-c747-11e9-975c-b3c7e2ab813c.png)

I then change projection of the map from WGS84 to North Pole Stereographic. It appears from this simple test that the only layer not reprojected is the terrain raster.  The gif image seems to be reprojected quite a bit off from where it should be according to the baseline polygon layer. It looks like only the corners of the image are reprojected. 

Rasters not reprojecting is something I posted about before and was looking into experimenting with GDAL but it fell lower in my to-do list.

![image](https://user-images.githubusercontent.com/5298460/63655014-1119ff80-c748-11e9-956f-376964f12782.png)

**Comment by sindizzy** [comment]
Here is how that looks in ArcGIS Pro for comparison

![image](https://user-images.githubusercontent.com/5298460/64071575-afbfc800-cc42-11e9-9cc5-cdb757e4ea50.png)

**Comment by VectorZita** [comment]
Yes, ArcGIS most likely performs a full warping of the raster, instead of simply reprojecting the corners (which is more like what the current implementation of the MapImageLayer does). The current implementation uses the `RasterBoundDataset`'s method to reproject, which ends up transforming the four corners of the image and adjusting it on these coordinates. This is not correct, of course, as your test result shows.
